### PR TITLE
Add a CloudStack builder

### PR DESCRIPTION
**Pull request by svanharmelen**
This PR adds a CloudStack builder, complete with tests and docs.

### CODE REVIEW

**Comment by svanharmelen** [comment]
@mitchellh I know that a few years ago another PR also offering a CloudStack builder was denied because of the inability for the Packer team to test the builder.

I guess that is probably still the case, but I would like to offer my help to maintain this builder in somewhat the same way I do for the Terraform project. Although I don't expect to be working on the core parts of Packer much, I can and would like to make sure this builder is properly maintained.

**Comment by rickard-von-essen** [comment]
:+1:

**Comment by svanharmelen** [comment]
@mitchellh and/or @rickard-von-essen, would be great to get some feedback this one in order to see if we can get this merged. Thx!

**Comment by rickard-von-essen** [comment]
What's the simplest way of getting a CloudStack up and running to test on? Is there anything like OpenStacks DevStack?

**Comment by svanharmelen** [comment]
@rickard-von-essen I guess this would be the easiest way to get something working: https://github.com/MissionCriticalCloud/bubble-cookbook

**Comment by svanharmelen** [comment]
@rickard-von-essen I think I may have found an even easier solution :wink: I just discussed if we could give HashiCorp access to a hosted bubble in our cloud, which seems to be no problem.

So let me see if I can set that up tomorrow and get back to you on this one. I it works out I will try to add some acceptance tests which we can then use to verify that stuff works as expected.

**Comment by rickard-von-essen** [comment]
@svanharmelen Sounds great! Just some notes:

1) I'm not a HashiCorp employee, I'm a community maintainer.
2) I could do some basic testing and review of the code and the docs, but to get this merged HashiCorp need to do thumbs up. The new HashiCorp Packer is @mwhooker.

If you need to send me any secrets you use GPG and my public key from [keybase.io/rickard_v_essen](https://keybase.io/rickard_v_essen) which is verified to be linked to my GitHub account, see [Gist Keybase proof - rickard_v_essen](https://gist.github.com/rickard-von-essen/601a50f023b1dc79160c)

**Comment by svanharmelen** [comment]
@rickard-von-essen it seems that it's a little more work to setup something that I can get approved to share with HashiCorp thenI expected. So not sure when that will happen as I am not doing that myself...

@mitchellh can you help out here by giving a GO/NO-GO based on my initial proposal (merge this PR with me offering to maintain the builder just like I do with the TF provider). If your OK with that, then we can move on. If not we need to look into what needs to happen before it can be merged.

Thanks!

File: website/source/docs/builders/cloudstack.html.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+-   `service_offering` (string) - The name or ID of the service offering used
+    for the instance.
+
+-   `soure_iso` (string) - The name or ID of an ISO that will be mounted before
+    booting the instance. This option is mutual exclusive with `source_template`.
+
+-   `source_template` (string) - The name or ID of the template used as base
+    template for the instance. This option is mutual explusive with `source_iso`.
+
+-   `template_name` (string) - The name of the new template. Defaults to a
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: Same as `instance_name`

File: website/source/docs/builders/cloudstack.html.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+-   `template_os` (string) - The name or ID of the template OS for the new
+    template that will be created.
+
+-   `zone` (string) - The name or ID of the zone where the instance will be
+    created.
+
+### Optional:
+
+-   `async_timeout` (int) - The time (in seconds) to wait for async calls to
+    finish. Defaults to 1800 seconds.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: It's nicer to use `time.Duration` so users can use "30m" instead of 1800. For a example see [amazon/common/run_config.go](https://github.com/mitchellh/packer/blob/master/builder/amazon/common/run_config.go#L36)

File: website/source/docs/builders/cloudstack.html.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+-   `http_get_only` (boolean) - Some cloud providers only allow HTTP GET calls to
+    their CloudStack API. If using such a provider, you need to set this to `true`
+    in order for the provider to only make GET calls and no POST calls.
+
+-   `hypervisor` (string) - The target hypervisor for the new template. This
+    option is required when using `source_iso`.
+
+-   `keypair` (string) - The name of the SSH key pair that will be used to
+    access the instance.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: This doesn't seems to be implemented yet.

File: website/source/docs/builders/cloudstack.html.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+-   `project` (string) - The name or ID of the project to deploy the instance to.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: Default value? Maybe it's obvious if you know CS.

File: website/source/docs/builders/cloudstack.html.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    is featured. Defaults to `false`.
+
+-   `template_public` (boolean) - Set to `true` to indicate that the template is
+    available for all accounts. Defaults to `false`.
+
+-   `template_password_enabled` (boolean) - Set to `true` to indicate the template
+    should be password enabled. Defaults to `false`.
+
+-   `template_requires_hvm` (boolean) - Set to `true` to indicate the template
+    requires hardware. Defaults to `false`.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: > Set to `true` to indicate the template requires hardware.

Probably should be _Hardware-assisted Virtual Machine_

File: website/source/docs/builders/cloudstack.html.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+-   `disk_size` (int) - The size (in GB) of the root disk of the new instance.
+    This option is required when using `source_template`.
+
+-   `http_get_only` (boolean) - Some cloud providers only allow HTTP GET calls to
+    their CloudStack API. If using such a provider, you need to set this to `true`
+    in order for the provider to only make GET calls and no POST calls.
+
+-   `hypervisor` (string) - The target hypervisor for the new template. This
+    option is required when using `source_iso`.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: Maybe name the most common values here.

**Comment by rickard-von-essen** [comment]
Overall this looks really good! Could you go through my comments.

File: website/source/docs/builders/cloudstack.html.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  "secret_key": "YOUR_SECRET_KEY",
+  
+  "disk_offering": "Small - 20GB",
+  "cidr_list": ["0.0.0.0/0"]
+  "hypervisor": "KVM",
+  "network": "management",
+  "service_offering": "small",
+  "source_iso": "CentOS-7.0-1406-x86_64-Minimal",
+  "zone": "NL1",
+
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: And here.

File: website/source/docs/builders/cloudstack.html.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+Here is a basic example.
+
+``` {.javascript}
+{
+  "type": "cloudstack",
+  "api_url": "https://cloudstack.company.com/client/api",
+  "api_key": "YOUR_API_KEY",
+  "secret_key": "YOUR_SECRET_KEY",
+
```
Comment (rickard-von-essen) [inline website/source/docs/builders/cloudstack.html.md]: Could you drop your trailing spaces here.

**Comment by svanharmelen** [comment]
@rickard-von-essen thanks for the review! I'm a little overloaded at the moment, but I'll cycle back with an updated PR somewhere this week (if all goes according to plan :wink:)

**Comment by mwhooker** [comment]
I just added this to the v0.12 milestone. 0.12 will be the last release before 1.0. After 1.0, features like this will be lower priority for a while. I'd really like to get this in by 0.12, or it will have to wait until next year.

**Comment by svanharmelen** [comment]
@mwhooker sounds good! I had a really busy week untill now, but will make sure I update the PR this week...

Would be a shame having to wait until next year for this one, so let's make some short itteration (if needed) and get this is :wink:

File: builder/cloudstack/builder.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	state.Put("config", b.config)
+	state.Put("hook", hook)
+	state.Put("ui", ui)
+
+	// Build the steps.
+	steps := []multistep.Step{
+		// &stepKeyPair{
+		// 	Debug:        b.config.PackerDebug,
+		// 	DebugKeyPath: fmt.Sprintf("cs_%s.pem", b.config.PackerBuildName),
+		// },
```
Comment (svanharmelen) [inline builder/cloudstack/builder.go]: Removed it... It's not in the docs, so nothing to update there.

File: builder/cloudstack/config.go [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+	Hypervisor        string   `mapstructure:"hypervisor"`
+	InstanceName      string   `mapstructure:"instance_name"`
+	Keypair           string   `mapstructure:"keypair"`
+	Network           string   `mapstructure:"network"`
+	Project           string   `mapstructure:"project"`
+	PublicIPAddress   string   `mapstructure:"public_ip_address"`
+	ServiceOffering   string   `mapstructure:"service_offering"`
+	SourceTemplate    string   `mapstructure:"source_template"`
+	SourceISO         string   `mapstructure:"source_iso"`
+	UserData          string   `mapstructure:"user_data"`
```
Comment (svanharmelen) [inline builder/cloudstack/config.go]: Added...

File: builder/cloudstack/config.go [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+	var errs *packer.MultiError
+
+	// Set some defaults.
+	if c.AsyncTimeout == 0 {
+		c.AsyncTimeout = 1800
+	}
+
+	if c.Comm.SSHUsername == "" {
+		c.Comm.SSHUsername = "root"
```
Comment (svanharmelen) [inline builder/cloudstack/config.go]: I fully agree, but don't want to make that change in this PR. I would say a change like that deserves it's own PR... Agree?

File: builder/cloudstack/ssh.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+	return config.hostAddress, nil
+}
+
+func sshConfig(state multistep.StateBag) (*ssh.ClientConfig, error) {
+	config := state.Get("config").(*Config)
+
+	clientConfig := &ssh.ClientConfig{
+		User: config.Comm.SSHUsername,
+		Auth: []ssh.AuthMethod{ssh.Password(config.Comm.SSHPassword)},
```
Comment (svanharmelen) [inline builder/cloudstack/ssh.go]: Added...

File: builder/cloudstack/step_create_template.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+func (s *stepCreateTemplate) Run(state multistep.StateBag) multistep.StepAction {
+	client := state.Get("client").(*cloudstack.CloudStackClient)
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	ui.Say(fmt.Sprintf("Creating template: %s", config.TemplateName))
+
+	// Retrieve the instance ID from the previously saved state.
+	instanceID, ok := state.Get("instance_id").(string)
```
Comment (svanharmelen) [inline builder/cloudstack/step_create_template.go]: But it doesn't hurt either. I prefer an error over a panic every day :wink:

File: builder/cloudstack/step_shutdown_instance.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+func (s *stepShutdownInstance) Run(state multistep.StateBag) multistep.StepAction {
+	client := state.Get("client").(*cloudstack.CloudStackClient)
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	ui.Say("Shutting down instance...")
+
+	// Retrieve the instance ID from the previously saved state.
+	instanceID, ok := state.Get("instance_id").(string)
```
Comment (svanharmelen) [inline builder/cloudstack/step_shutdown_instance.go]: Also same as above 😁

File: website/source/docs/builders/cloudstack.html.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+segmented below into two categories: required and optional parameters. Within
+each category, the available configuration keys are alphabetized.
+
+In addition to the options listed here, a
+[communicator](/docs/templates/communicator.html) can be configured for this
+builder.
+
+### Required:
+
+-   `api_url` (string) - The CloudStack API endpoint we will connect to.
```
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Nope, there aren't for CloudStack.

File: website/source/docs/builders/cloudstack.html.md (hunk 7 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Done...

File: website/source/docs/builders/cloudstack.html.md (hunk 8 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Done...

File: website/source/docs/builders/cloudstack.html.md (hunk 1 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Done...

File: website/source/docs/builders/cloudstack.html.md (hunk 6 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Done...

File: website/source/docs/builders/cloudstack.html.md (hunk 4 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Nope, no default... Working in a project is completely optional. So no project is also fine.

File: website/source/docs/builders/cloudstack.html.md (hunk 5 above)
Comment (svanharmelen) [inline website/source/docs/builders/cloudstack.html.md]: Updated...

**Comment by svanharmelen** [comment]
@rickard-von-essen @mwhooker just updated the PR... Let me know any remaining issues/comments so we can fix them. Thx!

File: builder/cloudstack/config.go [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+
+// NewConfig parses and validates the given config.
+func NewConfig(raws ...interface{}) (*Config, error) {
+	c := new(Config)
+	err := config.Decode(c, &config.DecodeOpts{
+		Interpolate:        true,
+		InterpolateContext: &c.ctx,
+		InterpolateFilter: &interpolate.RenderFilter{
+			Exclude: []string{
+				"run_command",
```
Comment (rickard-von-essen) [inline builder/cloudstack/config.go]: Sorry, you exclude `run_command` which is a non existing config option.

File: builder/cloudstack/step_create_template.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+	template, err := client.Template.CreateTemplate(p)
+	if err != nil {
+		ui.Error(fmt.Sprintf("Error creating the new template %s: %s", config.TemplateName, err))
+		return multistep.ActionHalt
+	}
+
+	// This is kind of nasty, but it appears to be needed to prevent corrupt templates.
+	// When CloudStack says the template creation is done and you then delete the source
+	// volume shortly after, it seems to corrupt the newly created template. Giving it an
+	// additional 30 seconds to really finish up, seem to prevent that from happening.
```
Comment (rickard-von-essen) [inline builder/cloudstack/step_create_template.go]: I think a 30s stop would make me wounder what's going on. It's not super important.

**Comment by rickard-von-essen** [comment]
👍 I'm happy!
@mwhooker handing this over to you to merge when you see fit.

File: builder/cloudstack/step_create_template.go (hunk 17 above)
Comment (svanharmelen) [inline builder/cloudstack/step_create_template.go]: I see... But you will not notice that, as the last output will be `Creating template...` which takes at least a couple of minutes to complete. So adding those 30 seconds will not be noticeable as you already know your have to wait for a few minutes until the template is copied/created.
